XenModel/Actions/CustomFields/RemoveCustomFieldAction.cs

```diff
@@ -0,0 +1,64 @@
+﻿/* Copyright (c) Cloud Software Group, Inc.
+ * 
+ * Redistribution and use in source and binary forms, 
+ * with or without modification, are permitted provided 
+ * that the following conditions are met: 
+ * 
+ * *   Redistributions of source code must retain the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer. 
+ * *   Redistributions in binary form must reproduce the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer in the documentation and/or other 
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using XenAdmin.CustomFields;
+using XenAdmin.Network;
+
+namespace XenAdmin.Actions
+{
+    /// <summary>
+    /// Removes a custom field definition from the pool.
+    /// </summary>
+    public class RemoveCustomFieldAction : AsyncAction
+    {
+        private readonly CustomFieldDefinition _definition;
+
+        public RemoveCustomFieldAction(IXenConnection connection, CustomFieldDefinition definition)
+            : base(connection, string.Format(Messages.DELETE_CUSTOM_FIELD, definition.Name),
+                string.Format(Messages.DELETING_CUSTOM_FIELD, definition.Name), false)
+        {
+            _definition = definition;
+
+            ApiMethodsToRoleCheck.AddRange("pool.add_to_gui_config", "pool.remove_from_gui_config",
+                "pool.remove_from_other_config",
+                "host.remove_from_other_config",
+                "VM.remove_from_other_config",
+                "SR.remove_from_other_config",
+                "VDI.remove_from_other_config",
+                "Network.remove_from_other_config");

```

Comment:
There are different permissions per object, for example, `pool.remove_from_other_config` is available to pool operators, while `vm.remove_from_other_config` is available to vm admins. Since the custom fields apply to all objects, I suppose we could have just used the highest call's RBAC, but I listed them all for completion.

To complicate things further, the permissions for actions adding/removing keys from dictionary properties depend sometimes on the specific key (which is why we have the method RbacMethodList.AddWithKey). For the objects' `other_config` we can't apply permissions per key because the latter is user defined. For the pool's `gui_config` we do know the key,  but I didn't add it because in another one or two actions that use the gui_config it wasn't taken into account and it wasn't clear to me whether this was ommission or on purpose, so I just copied the logic.

Reply:
Actually, SaveCustomFieldsAction does chek RBAC per key, although I'm not sure if this is useful since the key is not known - unless xapi check only for the prefix `XenCenter.CustomFields.` of the key.

Reply:
> There are different permissions per object, for example, pool.remove_from_other_config is available to pool operators, while vm.remove_from_other_config is available to vm admins. Since the custom fields apply to all objects, **I suppose we could have just used the highest call's RBAC, but I listed them all for completion**.

I hadn't thought of this. I'm happy with listing all in that case